cli/src/main/java/com/box/l10n/mojito/cli/command/checks/GlossaryTermTrie.java

```diff
@@ -0,0 +1,94 @@
+package com.box.l10n.mojito.cli.command.checks;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class GlossaryTermTrie {
+
+    private class Node {
+        Map<String, Node> children = new HashMap<>();
+        List<GlossaryTerm> glossaryTerms = new ArrayList<>();
+    }
+
+    private Node rootNode = new Node();
+
+    public GlossaryTermTrie(List<GlossaryTerm> terms) {
+        for(GlossaryTerm term: terms) {
+            this.add(term);
+        }
+    }
+
+    public void add(GlossaryTerm term) {
+        String[] words = term.getTerm().split("\\W+");
+        Node currentNode = rootNode;
+        for (String word : words) {
+            word = word.toLowerCase();
+            if(!currentNode.children.containsKey(word)) {
+                currentNode.children.put(word, new Node());
+            }
+            currentNode = currentNode.children.get(word);
+        }
+        currentNode.glossaryTerms.add(term);
+    }
+
+    public GlossarySearchResult runGlossaryCheck(String source) {
+        String[] words = source.split("\\W+");
+        GlossarySearchResult result = new GlossarySearchResult(source);
+        List<String> failures = new ArrayList<>();
+        for(int i = 0; i < words.length; i++) {
+            Node current = rootNode;
+            for (int j = i; j < words.length; j++) {
+                if (!current.children.containsKey(words[j].toLowerCase())){
+                    break;
+                }
+                current = current.children.get(words[j].toLowerCase());
+                if(!current.glossaryTerms.isEmpty()){
+                    checkGlossaryTerms(source, result, failures, current);
+                }
+            }
+        }
+        result.failures = failures;
+        return result;
+    }
+
+    private void checkGlossaryTerms(String source, GlossarySearchResult result, List<String> failures, Node current) {
+        String trimmedWhitespace = source.trim().replaceAll(" +", " ");
+        boolean allFailed = true;
```

Comment:
is the meaning here is "has match"? what has failed in this case? failed to find an exact match even though there was a case insensitive match?

Reply:
Yes failed to find an exact match even though there was a case insensitive match which means we have a failure